Mainnet/HashBattle/HashBattle/Arena.cs

```diff
@@ -0,0 +1,257 @@
+﻿using Stratis.SmartContracts;
+using System;
+using System.Text;
+
+/// <summary>
+/// A Stratis smart contract for running a game battle where owner will start the battle and maximum 4 users can enter a battle
+/// </summary>
+public class Arena : SmartContract
+{
+    public Arena(ISmartContractState smartContractState)
+    : base(smartContractState)
+    {
+        BattleOwner = Message.Sender;
+        NextBattleId = 1;
+    }
+
+    /// <summary>
+    /// Set the address deploying the contract as battle owner
+    /// </summary>
+    private Address BattleOwner
+    {
+        get => PersistentState.GetAddress(nameof(BattleOwner));
+        set => PersistentState.SetAddress(nameof(BattleOwner), value);
+    }
+
+    public Address PendingBattleOwner
+    {
+        get => PersistentState.GetAddress(nameof(PendingBattleOwner));
+        private set => PersistentState.SetAddress(nameof(PendingBattleOwner), value);
+    }
+
+    public void SetPendingBattleOwnership(Address pendingBattleOwner)
+    {
+        Assert(Message.Sender == BattleOwner, "UNAUTHORIZED");
+
+        PendingBattleOwner = pendingBattleOwner;
+
+        Log(new SetPendingDeployerOwnershipLog { From = Message.Sender, To = pendingBattleOwner });
+    }
+
+    public void ClaimPendingbattleOwnership()
+    {
+        var pendingBattleOwner = PendingBattleOwner;
+
+        Assert(Message.Sender == pendingBattleOwner, "UNAUTHORIZED");
+
+        var oldOwner = BattleOwner;
+
+        BattleOwner = pendingBattleOwner;
+        PendingBattleOwner = Address.Zero;
+
+        Log(new ClaimPendingDeployerOwnershipLog { From = oldOwner, To = pendingBattleOwner });
+    }
+
+    /// <summary>
+    /// Set the unique battleid of each battle
+    /// </summary>
+    public ulong NextBattleId
+    {
+        get => PersistentState.GetUInt64(ArenaStateKeys.NextBattleId);
+        private set => PersistentState.SetUInt64(ArenaStateKeys.NextBattleId, value);
+    }
+
+    /// <summary>
+    /// Battle owner will start the battle
+    /// </summary>
+    public ulong StartBattle(ulong fee)
+    {
+        Assert(Message.Sender == BattleOwner, "Only battle owner can start game.");
```

Comment:
Avoid overflows when you process price at the end.

```suggestion
               Assert(fee < ulong.MaxValue / MaxUsers, "Fee is too high");
```